src/other/metadata.jl

```diff
@@ -0,0 +1,117 @@
+# Defining behavior for DataFrames metadata
+abstract type AbstractMetaData end
+
+mutable struct MetaData <: AbstractMetaData
+    columndata::Dict{Symbol, Vector}
+end
+    
+MetaData() = MetaData(Dict{Symbol,Vector}())
+
+Base.isequal(x::MetaData, y::MetaData) = isequal(x.columndata, y.columndata)
+Base.:(==)(x::MetaData, y::MetaData) = isequal(x, y)
+
+Base.copy(x::MetaData) = MetaData(copy(x.columndata))
+Base.deepcopy(x::MetaData) = MetaData(copy(x.columndata)) # field is immutable
+
+function Base.getindex(x::MetaData, col_inds::AbstractVector)
+    new_columndata = copy(x.columndata)
+    for key in keys(new_columndata)
+        new_columndata[key] = new_columndata[key][col_inds]
+    end 
+    MetaData(new_columndata)
+end
+
+function Base.permute!(x::MetaData, p::AbstractVector)
+    for key in keys(x.columndata)
+        x.columndata[key] = permute!(x.columndata[key], p)
+    end 
+    nothing
+end
+
+function Base.permute(x::MetaData, p::AbstractVector)
+    new_metadata = copy(x)
+    permute!(new_metadata, p)
+end
+
+function newfield!(x::MetaData, ncol::Int, field::Symbol,)
+    x.columndata[field] = Vector{Any}([nothing for i in 1:ncol]) 
```

Comment:
Rather than using `Any`, I think `Union{eltype(info), Nothing}` would be more appropriate. That would be more efficient, and it's probably flexible enough for meta-data. We can always add API to choose a different type later if it turns out to be useful.